### PR TITLE
First non-trained version of Alpha Zero Dom player

### PR DESCRIPTION
**Pull request by AntiD2ta**
- [x] Deep convolutional neural network model
- [x] Selector
- [x] Rollout updated
- [x] Player updated
- [x] Some post-review errors fixed
- [x] Trainer

This PR includes some work on AZ trainer. This can be reviewed but is WIP

### CODE REVIEW

File: src/models/alpha_zero_model.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        assert isinstance(state_lists, list)
+        batch_size = len(state_lists)
+        batch = np.zeros((batch_size,) + STATE_SHAPE, dtype=np.float32)
+        for idx, state in enumerate(state_lists):
+            batch[idx] = state_to_list(state, 2542)
+        return torch.tensor(batch).to(self.device)
+
+    def valids_actions_to_tensor(self, valids_actions):
+        mask = state_to_list(valids_actions, 111)
```
Comment (e1Ru1o) [inline src/models/alpha_zero_model.py:164]: I think that we need to parameterize this number (111) in some way

File: src/training/alpha_zero_trainer.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        game_over = False
+        root = True
+        players = [BasePlayer(i) for i in range(4)]
+        manager = DominoManager()
+        manager.init(players, hand_out, self.max_number, self.pieces_per_player)
+
+        while not game_over:
+            stats = {}
+            cur_player = players[manager.domino.current_player]
+            selector = selector_maker(stats, cur_player.valid_moves(), cur_player.pieces_per_player - len(cur_player.pieces), root, 6)
```
Comment (e1Ru1o) [inline src/training/alpha_zero_trainer.py:81]: also we should parameterize this number(6)

File: src/training/alpha_zero_trainer.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            encoder = encoder_generator(self.max_number)
+            rollout = rollout_maker(stats, self.net)
+
+            if root:
+                root = False
```
Comment (e1Ru1o) [inline src/training/alpha_zero_trainer.py:86]:
```suggestion
            root = False
```

File: src/training/alpha_zero_trainer.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            state, action, pi = monte_carlo(
+                cur_player, 
+                encoder, 
+                rollout, 
+                selector,
+                handouts,
+                rollouts,
+            )
+            _, mask = get_valids_data(manager.domino)
+            game_over = manager.step(action=action)
```
Comment (e1Ru1o) [inline src/training/alpha_zero_trainer.py:97]:
```suggestion
            game_over = manager.step(True, action)
```

Comment (AntiD2ta) [reply src/training/alpha_zero_trainer.py:97]: That was the only fix I forgot

File: src/models/alpha_zero_model.py (hunk 1 above)
Comment (AntiD2ta) [reply src/models/alpha_zero_model.py:270]: That's a hard one. The same applies to 2542 above. These are numbers tied to STATE_SHAPE hardcoded value and policy_shape arg of init method. But are also tied to encoder implementation. I can set 111 and 2542 to be tied to STATE_SHAPE and policy_shape, but if these values change, then I think the encoder can't do anything about it and still will be encoding state as original STATE_SHAPE.

File: src/training/alpha_zero_trainer.py (hunk 2 above)
Comment (AntiD2ta) [reply src/training/alpha_zero_trainer.py:81]: Yes, that can be done

**Review by e1Ru1o — APPROVED** [review]

File: src/models/alpha_zero_model.py (hunk 1 above)
Comment (e1Ru1o) [reply src/models/alpha_zero_model.py:270]: let's skip this for now

**Pull request merged**